tests/conftest.py

```diff
@@ -57,10 +67,130 @@ def swarm_mode():
+        docker_client.ctr_mgr = "docker"
+        ctr_clients.append(docker_client)
+        try:
+            subprocess.check_call(
+                ["podman", "--version"],
+                stdout=subprocess.DEVNULL,
+                stderr=subprocess.DEVNULL,
+            )
+        except (subprocess.CalledProcessError, FileNotFoundError):
+            logger.warning("Unable to parameterise the tests with podman")
+        else:
+            podman_client = DockerClient(client_call=["podman"])
+            podman_client.ctr_mgr = "podman"
+            ctr_clients.append(podman_client)
+    else:
+        unsupported_ctr_exes: List[str] = []
+        for exe in ctr_exes:
+            client = DockerClient(client_call=[exe])
+            # Check the client is available by running '<client> --version'.
+            try:
+                version_output = subprocess.check_output([exe, "--version"], text=True)
+            except (subprocess.CalledProcessError, FileNotFoundError):
+                unsupported_ctr_exes.append(exe)
+                continue
+            # Check the daemon is running (if required) with '<client> version'.
+            try:
+                subprocess.check_call([exe, "version"], stdout=subprocess.DEVNULL)
+            except subprocess.CalledProcessError:
+                unsupported_ctr_exes.append(exe)
+                continue
+
+            ctr_clients.append(client)
+
+            # Deduce whether docker or podman from version output.
+            if "docker" in version_output.lower():
+                client.ctr_mgr = "docker"
+            elif "podman" in version_output.lower():
+                client.ctr_mgr = "podman"
+            else:
+                logger.warning(
+                    f"Unrecognised container manager from '{exe} --version', "
+                    "assuming docker"
+                )
+                client.ctr_mgr = "docker"
+
+        if unsupported_ctr_exes:
+            raise Exception(
+                "'<ctr_exe> --version' failed for the following specified exes: "
+                + ",".join(unsupported_ctr_exes)
+            )
+
+    return ctr_clients
+
+
+def pytest_configure(config: pytest.Config) -> None:
+    # Filter warnings
+    if not pydantic.__version__.startswith("1"):
+        config.addinivalue_line(
+            "filterwarnings", "error::pydantic.PydanticDeprecatedSince20"
+        )
+    # Register markers
+    config.addinivalue_line(
+        "markers",
+        "ctr_mgr(MGR, MGR, ...): Container managers supported by the test",
+    )
+    # Determine supported container managers.
+    config.ctr_clients = _get_ctr_clients(config.getoption("--ctr-exe"))
+
+
+def pytest_generate_tests(metafunc: pytest.Metafunc) -> None:
+    if "ctr_client" in metafunc.fixturenames:
+        metafunc.parametrize(
+            "ctr_client",
+            metafunc.config.ctr_clients,
+            ids=[
+                client.client_config.client_call[0]

```

Comment:
If there are multiple args, it may be better to do `"-".join(client.client_config.client_call)` no? It's just to be as general as possible.

Reply:
`--ctr-exe` currently only supports specifying the exe, not a list of args, but could still change this

Reply:
no worries, i'm just nitpicking here, it's not a big deal